include/deal.II/lac/solver_gmres.h

```diff
@@ -710,81 +723,372 @@ SolverGMRES<VectorType>::givens_rotation(Vector<double> &h,
 
 
 
-template <class VectorType>
-inline double
-SolverGMRES<VectorType>::iterated_modified_gram_schmidt(
-  const internal::SolverGMRESImplementation::TmpVectors<VectorType>
-    &                                       orthogonal_vectors,
-  const unsigned int                        dim,
-  const unsigned int                        accumulated_iterations,
-  VectorType &                              vv,
-  Vector<double> &                          h,
-  bool &                                    reorthogonalize,
-  const boost::signals2::signal<void(int)> &reorthogonalize_signal)
+namespace internal
 {
-  Assert(dim > 0, ExcInternalError());
-  const unsigned int inner_iteration = dim - 1;
+  namespace SolverGMRESImplementation
+  {
+    template <class VectorType>
+    void
+    Tvmult_add(const unsigned int dim,
+               const VectorType & vv,
+               const internal::SolverGMRESImplementation::TmpVectors<VectorType>
+                 &             orthogonal_vectors,
+               Vector<double> &h)
+    {
+      for (unsigned int i = 0; i < dim; ++i)
+        h[i] += vv * orthogonal_vectors[i];
+    }
+
+
+
+    template <class Number>
+    void
+    Tvmult_add(
+      const unsigned int                                                   dim,
+      const LinearAlgebra::distributed::Vector<Number, MemorySpace::Host> &vv,
+      const internal::SolverGMRESImplementation::TmpVectors<
+        LinearAlgebra::distributed::Vector<Number, MemorySpace::Host>>
+        &             orthogonal_vectors,
+      Vector<double> &h)
+    {
+      unsigned int j = 0;
+
+      if (dim <= 128)
+        {
```

Comment:
If I read the code correctly, you did not include an `else` clause? Anyway, when I see this code, I think we are nearly there already. Can we just introduce a loop like
```suggestion
      for (unsigned int batch = 0; batch < (dim + 127) / 128; ++batch)
        {
          const unsigned int local_dim = std::min(128U, dim - batch * 128);
```
replace all loops to `dim` by `local_dim` and finally write access to `h(i)` as `h(i + batch * 128)` as well as `orthogonal_vec[i + batch * 128]`.

Reply:
It has an implicit "else"-statement. The counter: https://github.com/dealii/dealii/blob/842ee4537c3152353c144628643af79e6a1ab79c/include/deal.II/lac/solver_gmres.h#L754 is not incremented and stays at 0. However, I can modify it as you have suggested.

Reply:
Ah, I see now, I missed that. I let you decide what you prefer; if staying with the present code I would suggest to add a comment to avoid others make my mistake.

Reply:
I have added a comment!